api/src/test/java/org/openmrs/api/VisitServiceTest.java

```diff
@@ -963,32 +965,74 @@ public void saveVisit_shouldAssociateEncounterWithVisitOnSaveEncounter() throws
 	 * @see VisitService#saveVisit(Visit)
 	 */
 	@Test
-	@Verifies(value = "save visit should persist new encounter", method = "saveVisit(Visit)")
-	public void saveVisit_shouldPersistNewEncounter() throws Exception {
+	@Verifies(value = "save visit should not persist new encounter", method = "saveVisit(Visit)")
+	public void saveVisit_shouldNotPersistNewEncounter() throws Exception {
 		
 		VisitService vs = Context.getVisitService();
 		Visit visit = vs.getVisit(1);
 		
 		Encounter encounter = new Encounter();
 		encounter.setEncounterDatetime(new Date());
+		encounter.setEncounterType(Context.getEncounterService().getEncounterType(1));
 		encounter.setPatient(visit.getPatient());
 		encounter.setLocation(visit.getLocation());
 		visit.addEncounter(encounter);
 		
 		vs.saveVisit(visit);
-		
+
+		assertNull(encounter.getEncounterId()); // should not have saved the encounter
+
+		Context.getEncounterService().saveEncounter(encounter); //save the encounter
+		Integer encounterId = encounter.getEncounterId();
+
 		Context.flushSession();
 		Context.clearSession();
-		
-		Integer encounterId = encounter.getEncounterId();
-		
+
 		// reload the visit
 		visit = Context.getVisitService().getVisit(1);
-		
+
 		assertEquals(1, visit.getEncounters().size());
 		assertEquals(encounterId, ((Encounter) visit.getEncounters().toArray()[0]).getEncounterId());
 	}
-	
+
+	/**
+	 * @see VisitService#saveVisit(Visit)
+	 */
+	@Test
+	@Verifies(value = "save visit should not update existing encounter", method = "saveVisit(Visit)")
+	public void saveVisit_shouldNotUpdateExistingEncounter() throws Exception {
+		executeDataSet(VISITS_WITH_DATES_XML);
+
+		VisitService vs = Context.getVisitService();
+		Visit visit = vs.getVisit(7);
+		Encounter encounter = visit.getNonVoidedEncounters().get(0);
+		assertEquals(0, encounter.getAllObs(true).size());
+		int obsBeforeCount = encounter.getAllObs(true).size();
+		Obs obs = new Obs();
+		obs.setConcept(Context.getConceptService().getConcept(19));
+		obs.setValueText("new obs");
+		encounter.addObs(obs);
+
+		vs.saveVisit(visit);
+
+		assertNull(encounter.getObs().iterator().next().getObsId()); // should not have saved the encounter
```

Comment:
What I'm saying is that if encounter.location is set to location A, change it to location B, save the visit and make sure that the encounter location is still set to location A and not B after the flush and clear